dataplane/saiserver/l2.go

```diff
@@ -16,48 +16,198 @@ package saiserver
+		// Add a member.
+		mg.groups[gid][mid] = m
+		attr := &saipb.L2McGroupMemberAttribute{
+			L2McGroupId:  &m.groupId,
+			L2McOutputId: &m.outputId,
+		}
+		mg.mgr.StoreAttributes(mid, attr)
+	}
+	mList := []uint64{}
+	for _, m := range mg.groups[gid] {
+		mList = append(mList, m.oid)
+	}
+	gAttr := &saipb.L2McGroupAttribute{}
+	if err := mg.mgr.PopulateAllAttributes(fmt.Sprint(gid), gAttr); err != nil {
+		return err
+	}
+	gAttr.L2McMemberList = mList
+	*gAttr.L2McOutputCount = uint32(len(mList))
+	mg.mgr.StoreAttributes(gid, gAttr)
+
+	var actions []*fwdpb.ActionDesc
+	for _, member := range mg.groups[gid] {
+		portId, err := mg.portNidFromBrirdgeId(ctx, member.outputId)
+		if err != nil {
+			return err
+		}
+		actions = append(actions, &fwdpb.ActionDesc{
+			ActionType: fwdpb.ActionType_ACTION_TYPE_MIRROR,
+			Action: &fwdpb.ActionDesc_Mirror{
+				Mirror: &fwdpb.MirrorActionDesc{
+					PortId: &fwdpb.PortId{
+						ObjectId: &fwdpb.ObjectId{
+							Id: fmt.Sprint(portId),
+						},
+					},
+				},
+			},
+		})
+	}
+	entries := &fwdpb.TableEntryAddRequest{
+		ContextId: &fwdpb.ContextId{Id: mg.dataplane.ID()},
+		TableId: &fwdpb.TableId{
+			ObjectId: &fwdpb.ObjectId{
+				Id: L2MCGroupTable,
+			},
+		},
+		Entries: []*fwdpb.TableEntryAddRequest_Entry{{
+			EntryDesc: &fwdpb.EntryDesc{
+				Entry: &fwdpb.EntryDesc_Exact{
+					Exact: &fwdpb.ExactEntryDesc{
+						Fields: []*fwdpb.PacketFieldBytes{{
+							Bytes: binary.BigEndian.AppendUint64(nil, gid),
+							FieldId: &fwdpb.PacketFieldId{
+								Field: &fwdpb.PacketField{
+									FieldNum: fwdpb.PacketFieldNum_PACKET_FIELD_NUM_L2MC_GROUP_ID,
+								},
+							},
+						}},
+					},
+				},
+			},
+			Actions: actions,
+		}},
+	}
+	_, err := mg.dataplane.TableEntryAdd(ctx, entries)
+	return err
+}
+
+func (mg *l2mcGroup) CreateL2McGroupMember(ctx context.Context, req *saipb.CreateL2McGroupMemberRequest) (*saipb.CreateL2McGroupMemberResponse, error) {
+	if mg.groups[req.GetL2McGroupId()] == nil {
+		return nil, status.Errorf(codes.FailedPrecondition, "cannot find L2MC group with group ID=%d", req.GetL2McGroupId())
+	}
+	if _, err := mg.portNidFromBrirdgeId(ctx, req.GetL2McOutputId()); err != nil {
+		return nil, err
+	}
 	id := mg.mgr.NextID()
+	if err := mg.updateGroupMember(ctx, req.GetL2McGroupId(), id, &l2mcGroupMember{
+		oid:      id,
+		outputId: req.GetL2McOutputId(),
+		groupId:  req.GetL2McGroupId(),
+	}); err != nil {
+		return nil, err
+	}
 	return &saipb.CreateL2McGroupMemberResponse{Oid: id}, nil
 }
 
-// TODO: Implement this.
-func (mg *l2mcGroup) RemoveL2McGroup(context.Context, *saipb.RemoveL2McGroupRequest) (*saipb.RemoveL2McGroupResponse, error) {
+func (mg *l2mcGroup) RemoveL2McGroup(ctx context.Context, req *saipb.RemoveL2McGroupRequest) (*saipb.RemoveL2McGroupResponse, error) {
+	if mg.groups[req.GetOid()] == nil {
+		return nil, status.Errorf(codes.FailedPrecondition, "cannot find L2MC group %q", req.GetOid())
+	}
+	// Remove all members in the group.
```

Comment:
i think it's simpler to just remove that entry from the table and delete the map rather than remove all the members

Reply:
done. Also add a unit test.